core/src/test/java/io/questdb/griffin/engine/functions/date/ExtractFunctionTest.java

```diff
@@ -0,0 +1,941 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.date;
+
+import io.questdb.griffin.AbstractGriffinTest;
+import org.junit.Test;
+
+public class ExtractFunctionTest extends AbstractGriffinTest {
+
+    @Test
+    public void test1997Millennium() throws Exception {
```

Comment:
The 0-th millenium (which includes important historical events) simply does not exist for questdb:

select extract(millennium from to_timestamp('999-12-31T23:59:59.999999Z')) -> null;

select extract(millennium from  to_timestamp('1000-01-01T00:00:00.000000Z')) -> 1

Reply:
to_timestamp expects ts to start with 4 year digits . For example 
```sql
select extract(millennium from to_timestamp('0999-12-31T23:59:59.999999Z')) 
```
produces 0999-12-31T23:59:59.999999Z

Reply:
It is but there are many more issues with to_timestamp() and it would be better to address them in separate PR .

Reply:
I would be keen to hear what are the issues. Leading zero is easy to fix

Reply:
Just a few examples . I bet there are more inconsistencies . 

```sql
select to_timestamp('9999-11-01T23:59:59.999999Z', 'yyyy-MM-ddTHH:mm:ss.SSSUUUZ')
9999-11-01T23:59:59.999999Z

select to_timestamp('999-11-01T23:59:59.999999Z', 'yyy-MM-ddTHH:mm:ss.SSSUUUZ')
0009-11-01T23:59:59.999999Z

select to_timestamp('99-11-01T23:59:59.999999Z', 'yy-MM-ddTHH:mm:ss.SSSUUUZ')
1999-11-01T23:59:59.999999Z

select to_timestamp('9-11-01T23:59:59.999999Z', 'y-MM-ddTHH:mm:ss.SSSUUUZ')
0009-11-01T23:59:59.999999Z
```

Reply:
```java
final DateTimeFormatter FMT = new DateTimeFormatterBuilder()
        .appendPattern("yyyy-MM-dd HH:mm:ss")
        .toFormatter()
        .withZone(ZoneId.of("Europe/London"));
Instant instant = FMT.parse("9999-02-15 18:35:24", Instant::from);

System.out.println(instant);
```

```
9999-02-15T17:35:24Z
```


```java
final DateTimeFormatter FMT = new DateTimeFormatterBuilder()
        .appendPattern("yyy-MM-dd HH:mm:ss")
        .toFormatter()
        .withZone(ZoneId.of("Europe/London"));
Instant instant = FMT.parse("999-02-15 18:35:24", Instant::from);

System.out.println(instant);
```

```
0999-02-15T18:36:39Z
```

Reply:
@marregui Java Date library will not parse three digit year as UTC. Try this:

```
String output = 
    Instant.parse ( "999-04-15T20:08:18Z" )
           .atZone ( ZoneId.of ( "America/Montreal" ) )
           .format ( 
               DateTimeFormatter.ofLocalizedDateTime ( FormatStyle.FULL )
                                .withLocale ( Locale.CANADA_FRENCH ) 
           )
;
```

Reply:
all of the inconsistencies fixed here: https://github.com/questdb/questdb/pull/2042/commits/4c94e48aeb774e1709ccefe4f65576dc072d4a8e

Reply:
Just for comparison - in PG if mask has less than 4 year chars it's interpreted as 2000 + year  :

```sql
postgres=# select to_timestamp('9-11-01T23:59:59', 'Y-MM-ddTHH:mi:SS');
      to_timestamp
------------------------
 2009-11-01 00:59:59+01
(1 row)


postgres=# select to_timestamp('19-11-01T23:59:59.123456', 'YY-MM-ddTHH:mi:SS.US');
         to_timestamp
-------------------------------
 2019-11-01 00:59:59.123456+01
(1 row)


postgres=# select to_timestamp('191-11-01T23:59:59.123456', 'YYY-MM-ddTHH:mi:SS.US');
         to_timestamp
-------------------------------
 2191-11-01 00:59:59.123456+01
(1 row)
```

In QDB handling of two digits doesn't look consistent with that of a single and three year digits :

```sql
select to_timestamp('0-11-01T23:59:59.999999Z', 'y-MM-ddTHH:mm:ss.SSSUUUZ')
0001-11-01T23:59:59.999999Z

select to_timestamp('9-11-01T23:59:59.999999Z', 'y-MM-ddTHH:mm:ss.SSSUUUZ')
0009-11-01T23:59:59.999999Z

select to_timestamp('29-11-01T23:59:59.999999Z', 'yy-MM-ddTHH:mm:ss.SSSUUUZ')
2029-11-01T23:59:59.999999Z

select to_timestamp('99-11-01T23:59:59.999999Z', 'yy-MM-ddTHH:mm:ss.SSSUUUZ')
2099-11-01T23:59:59.999999Z

select to_timestamp('029-11-01T23:59:59.999999Z', 'yyy-MM-ddTHH:mm:ss.SSSUUUZ')
0029-11-01T23:59:59.999999Z

select to_timestamp('929-11-01T23:59:59.999999Z', 'yyy-MM-ddTHH:mm:ss.SSSUUUZ')
0929-11-01T23:59:59.999999Z
```

Maybe we could make it work as in PG . 
By the way - to_timestamp format mask in PG is from QDB . 
I think it'd be good to make it clear in documentation because e.g. default  QDB mask doesn't work in PG : 

```sql
postgres=# select to_timestamp('1239-11-01T23:59:59.999999Z', 'yyyy-MM-ddTHH:mm:ss.SSSUUUZ')
postgres-# ;
ERROR:  conflicting values for "mm" field in formatting string
DETAIL:  This value contradicts a previous setting for the same field type.
```

In PG and Oracle minutes are specified with 'MI'.

Reply:
Aligning date format with PG is well outside of scope of this PR, also it is a lot of meticulous work.

The background of this work is to make QuestDB work with DataGrip, which entailed implementing `extract` syntax. We can perhaps deal with inconsistencies in separate PRs?